README
What is this?
=============

Chat is an activity used to exchange messages with friends or classmates.
You can chat about a topic you are studying or you can share something private that happened in your life.

How to use?
===========

Chat is part of the Sugar desktop.  Please refer to;

* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
* [How to use Sugar](https://help.sugarlabs.org/),
* [How to use Chat](https://help.sugarlabs.org/chat.html)


NEWS
86

* Replace TelepathyGLib operations with D-Bus (Aniket Mathur),
* Advise of connection progress (Swarup N),
* Revert "Promote activity tool bar" (James Cameron),
* Feature: search chat text (Swarup N),
* Port to Python 3 (Rahul Bothra),
* Remove empty locale files (James Cameron),
* flake8 fixes (James Cameron, Aniket Mathur),

85

* New translations (Chris Leonard et al),
* Updated POT file (James Cameron),
* flake8 fixes (James Cameron),
* Move activity tool bar into main tool bar (James Cameron),
* Fix emoji menu horizontal overflow sugarlabs/sugar#759 (James Cameron),
* Dismiss emoji menu on escape key (James Cameron),
* Decrease startup time by delaying emoji menu construction (James Cameron),
* Fix log warning on entry focus (James Cameron),
* Fix link clicking after GTK 3.19.9 (James Cameron),
* Port Telepathy static binding to PyGObject binding (Rahul Bothra),
* Add README source repository documentation (Rudra Sadhu),
* Update metadata (Vipul Gupta),

84

* New translations (Chris Leonard et al),
* Updated POT file (James Cameron),
* Fix version warnings from Python GI API,

83

* Fix issue with close (Quozl)
* New translations
* Add max_participants (AbrahamAB)

82

* New translations
* Added repository to activity.info

81

BUG FIX:
* Don't try to write smile files to readonly directory

80

* Add /me support
* Use a bit more of the roundbox width for the text message
* Fix problem with resizing textview for long messages (w/Gonzalo)
* Fix padding problem with OSK that pushed messages off screen
* Tighten up spacing to fit more messages on screen

79

* GTK3 update by Flavio
* Sound alerts by Ignacio
* Notifications by Jorge and SAMdroid
* Add more options (unicode chars) to smilies palette
* Use standard activity toolbar (works with portrait mode)
* Add filler for OSK

78

* Replace cjson with json #3142
* Update translations

77

* Fix issue of not packaging .mo translation files
* Chat text lines offset on screen lower than username #3634
* Update translations

76

* Support "Write to Journal Anytime" feature in Sugar 0.86

75

* Scroll the chat log #2782 (Manuel Kaufmann)
* Don't treat the whole word as an URL (Manuel Kaufmann)
* Update translations

74

* Remove use of hippo (Gonzalo Odiard)
* Update translations

73

* Return langs back by generating locale/ directory

72

* Update translations

71

* Chat i18n enhancement #2903 (Chris Leonard)
* Remove the KeepButton due to deprecation (Simon Schampijer)
* Fix View Source behaviour
* Update i18n

70

* Support gtk<2.18, eg, fc11
* Make smilies sensitive only in online mode
* After choosing a smyle, focus entry widget
* Update translations for de

69

* Add chat smilies (Mukesh Gupta)
* Update translations for en_US

68

* Update translations for de and ta

67

* Remove access to a private member (Tomeu Vizoso)
* Update translations for hi, nl

66

* Utilize new toolbars design
* Add new translations: mg, sq, ta

65

* Support new activity.info exec parameter (alsroot)
* #402: "share or invite" hint even on resume of shared instance (alsroot)

46

* Version bump to allow for stable releases after v48
* #8772: Fix journal entry creation in Chat for uri-list (kevix)
* #8471: Allow resuming Chat log in Write (morgs)
* #8411: Add license to activity.info (morgs)
* Remove parameter from bundlebuilder.start (morgs)
* Add update_url for software updater (morgs)

45

* Updated translations: sl, nb, el, mr, rw, ur, ne
* Fixed MANIFEST to include all translations (morgs)

44

* #7633: Close the text channel when stopping a 1-1 chat (morgs)
* #7717: Log incoming messages (morgs)
* #7692: Don't show pending messages when joining a chat (morgs)
* Updated translations: nl, te, es, mn

43

* Updated translations: zh_TW, ja

42

* #6036: Show timestamp as elapsed time instead of date (morgs)
* Updated translations: fr, mvo, pis, af, sd, pap, tpi, ar, de

41

* Updated translations: mr, de, ht, km, es, it
* #6036: Add separator after old chat history (morgs)
* #6298: Implement 1-1 private chat with non Sugar Jabber clients (morgs)

40

#5767: Use black text on light fill colors (matthias)

39

* ACK received messages (cassidy)
* Handle pending messages when setting the message handler (cassidy)

38

* Updated translations: zh_TW, de, it

37

* UI Change: Merge multiple sequential messages from same author (morgs)
* Updated translation: ar (pootle)
* #6561: Fix RTL message alignment (Arabic) (khaled)

36

* #5053: Reduce white space around boxes (morgs)
* #6621: set entry sensitive not editable (morgs)
* Add license to activity.py (morgs)
* #6743: border around gtk.entry (morgs)
* Reduce telepathy code based on improved PS channel creation API (morgs)
* Open URLs via show_object_in_journal (morgs)
* Update pippy metadata based on Pippy (morgs)
* Updated translations (pootle)

35

* #6066: Make web links copied to clipboard, pasteable in Write, Browse,
  Terminal (morgs)
* Added AUTHORS, COPYING (morgs)
* Updated translations (pootle)

34

* Updated translations: ur, bn (pootle)
* #2351: Scrolling fixed (marcopg)

33


32

* #5542: Repackaged as a Pippy application. (cscott)

31

* Updated translations: fa, is (pootle)
* #5080: Copy to clipboard with targets (morgs)

30

* Updated translations: es, fr, ne, pt, ro, ru, ur (pootle)
* #5160: Chat should not autoscroll while you scroll up (morgs)

29

* #5080: add a "copy to clipboard" palette for URL's (cassidy)
* Updated translations: fr, es, el, de, ar, zh_TW, it, nl, pt_BR (pootle)

28

* use NotifyAlert from sugar.graphics.alert instead of local
  copy (thanks erikos!) (morgs)

27

* Use sugar.graphics.alert to show status info (morgs)
* #4320: better URL handling and display (morgs)
* #4331: Don't crash/ignore non-Sugar buddies (morgs)

26

* #4320 Better URL support (morgs)

25

* #3417 Resuming shows chat history (morgs)
* self.set_title() considered harmful (morgs)
* New UI look per Eben's mockups (morgs)

24

* #3556: Updated spanish translation (morgs)

23


chat/smilies.py
# -*- coding: utf-8 -*-
# Copyright 2010, Mukesh Gupta
# Copyright 2010, Aleksey Lim
# Copyright 2014, Walter Bender
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

import os
from gettext import gettext as _
from gi.repository import GdkPixbuf
from sugar3.graphics import style
from sugar3.activity.activity import get_bundle_path

THEME = \
    [
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :-) and :)
        ('smile', _('Smile'), [':-)', ':)']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are ;-) and ;)
        ('wink', _('Winking'), [';-)', ';)']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :-/ and :/
        ('confused', _('Confused'), [':-/', ':/']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :-( and :(
        ('sad', _('Sad'), [':-(', ':(']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :-D and :D
        ('grin', _('Grin'), [':-D', ':D']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :-| and :|
        ('neutral', _('Neutral'), (':-|', ':|')),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :-O, :O, =-O and =O
        ('shock', _('Shock'), [':-O', ':O', '=-O', '=O']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are B-), B), 8-) and 8)
        ('cool', _('Cool'), ['B-)', 'B)', '8-)', '8)']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :-P and :P
        ('tongue', _('Tongue'), [':-P', ':P']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :">
        ('blush', _('Blushing'), [':">']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :'-( and :'(
        ('weep', _('Weeping'), [":'-(", ":'("]),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are O-), O), O:-) and O:)
        ('angel', _('Angel'), ['O-)', 'O)', 'O:-)', 'O:)']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :-$ and :-$
        ('shutup', _("Don't tell anyone"), (':-$', ':-$')),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are x-(, x(, X-( and x-(
        ('angry', _('Angry'), ('x-(', 'x(', 'X-(', 'x-(')),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are >:> and >:)
        ('devil', _('Devil'), ('>:>', '>:)')),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :-B, :B
        ('nerd', _('Nerd'), (':-B', ':B')),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :-*, :*
        ('kiss', _('Kiss'), (':-*', ':*')),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :))
        ('laugh', _('Laughing'), [':))']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are I-)
        ('sleep', _('Sleepy'), ['I-)']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are :-&
        ('sick', _('Sick'), [':-&']),
        # TRANS: A smiley (http://en.wikipedia.org/wiki/Smiley) explanation
        # TRANS: ASCII-art equivalents are /:)
        ('eyebrow', _('Raised eyebrows'), ['/:)']),
        ('unicode-1', _('Heart'), ['♥']),
        ('unicode-2', _('Airplane'), ['✈']),
        ('unicode-3', _('Music'), ['♬']),
        ('unicode-4', _('Check box'), ['☑']),
        ('unicode-5', _('Spade'), ['♠']),
        ('unicode-6', _('Telephone'), ['☎']),
        ('unicode-7', _('X box'), ['☒']),
        ('unicode-8', _('Cadusis'), ['☤']),
        ('unicode-9', _('Female'), ['♀']),
        ('unicode-10', _('Star'), ['✩']),
        ('unicode-11', _('Letter'), ['✉']),
        ('unicode-12', _('Poison'), ['☠']),
        ('unicode-13', _('Check mark'), ['✔']),
        ('unicode-14', _('Male'), ['♂']),
        ('unicode-15', _('Star'), ['★']),
        ('unicode-16', _('Shamrock'), ['☘']),
        ('unicode-17', _('Recycle'), ['♺']),
        ('unicode-18', _('X'), ['✖']),
        ('unicode-19', _('Fire'), ['♨']),
        ('unicode-20', _('Leaf'), ['❦']),
        ('unicode-21', _('Cloud'), ['☁']),
        ('unicode-22', _('V'), ['✌']),
        ('unicode-23', _('Crown'), ['♛']),
        ('unicode-24', _('Floret'), ['❁']),
        ('unicode-25', _('Star and crescent'), ['☪']),
        ('unicode-26', _('Umbrella'), ['☂']),
        ('unicode-27', _('Pencil'), ['✏']),
        ('unicode-28', _('Floret'), ['❀']),
        ('unicode-29', _('Snowman'), ['☃']),
        ('unicode-30', _('Point right'), ['☛']),
        ('unicode-31', _('Horse'), ['♞']),
        ('unicode-32', _('Floret'), ['✿']),
        ('unicode-33', _('Peace sign'), ['☮']),
        ('unicode-34', _('Sun'), ['☼']),
        ('unicode-35', _('Point left'), ['☚']),
        ('unicode-36', _('Floret'), ['✾']),
        ('unicode-37', _('Yin yang'), ['☯']),
        ('unicode-38', _('Moon'), ['☾']),
        ('unicode-39', _('Point up'), ['☝']),
        ('unicode-40', _('Floret'), ['✽']),
        ('unicode-41', _('Meteor'), ['☄']),
        ('unicode-42', _('Point down'), ['☟']),
        ('unicode-43', _('Floret'), ['✺']),
        ('unicode-44', _('Scissors'), ['✂']),
        ('unicode-45', _('Pen'), ['✍']),
        ('unicode-46', _('Floret'), ['✵']),
        ('unicode-47', _('Stars'), ['⁂']),
        ('unicode-48', _('Place'), ['⌘']),
        ('unicode-49', _('High voltage'), ['⚡']),
    ]

SMILIES_SIZE = int(style.STANDARD_ICON_SIZE * 0.75)
_catalog = {}


def _smiley_to_theme_name(smiley):
    for theme in THEME:
        if smiley in theme[2]:
            return theme[0]
    return None


def parse(text):
    '''Parse text and find smiles.
    :param text:
    string to parse for smilies
    :returns:
    array of string parts and pixbufs
    '''
    if not _catalog:
        init()

    result = [text]

    for smiley in sorted(list(_catalog.keys()), key=lambda x: - len(x)):
        new_result = []
        for word in result:
            if isinstance(word, GdkPixbuf.Pixbuf):
                new_result.append(word)
            else:
                parts = word.split(smiley)
                for i in parts[:-1]:
                    name = _smiley_to_theme_name(smiley)
                    new_result.append(i)
                    if name is not None and 'unicode' in name:
                        new_result.append(smiley)
                    else:
                        new_result.append(_catalog[smiley])
                new_result.append(parts[-1])
        result = new_result

    return result


def init():
    if _catalog:
        return

    svg_dir = os.path.join(get_bundle_path(), 'icons', 'smilies')

    for index, (name, hint, codes) in enumerate(THEME):
        archivo = os.path.join(svg_dir, '%s.svg' % (name))
        if name[0:7] == 'unicode':
            # Create the icon from unicode character on the fly
            pl = GdkPixbuf.PixbufLoader.new_with_type('svg')
            pl.write(_generate_svg(codes[0]))
            pl.close()
            pixbuf = pl.get_pixbuf()
            if not os.path.exists(archivo):
                try:
                    fd = open(archivo, 'w')
                    fd.write(_generate_svg(codes[0]))
                    fd.close()
                except IOError:
                    pass
        else:
            pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(
                archivo, SMILIES_SIZE, SMILIES_SIZE)
        for i in codes:
            _catalog[i] = pixbuf
            THEME[index] = (archivo, hint, codes)


def _generate_svg(letter):
    # TODO: Adjust font size and character positioning
    return ('<?xml version="1.0" encoding="UTF-8" standalone="no"?>\n' +
            '<svg\n' +
            ' xmlns:dc="http://purl.org/dc/elements/1.1/"\n' +
            ' xmlns:cc="http://creativecommons.org/ns#"\n' +
            ' xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"\n' +
            ' xmlns:svg="http://www.w3.org/2000/svg"\n' +
            ' xmlns="http://www.w3.org/2000/svg"\n' +
            ' version="1.0"\n' +
            ' width="55"\n' +
            ' height="55"\n' +
            '>\n' +
            '<text\n' +
            ' x="0"\n' +
            ' y="42"\n' +
            'style="font-size:40px;font-style:normal;font-weight:normal;\n' +
            'fill:#ffffff;fill-opacity:1;stroke:none;' +
            'font-family:Bitstream Vera Sans">\n' +
            '<tspan\n' +
            ' x="0"\n' +
            ' y="42">\n' +
            letter +
            '</tspan></text>\n' +
            '</svg>\n').encode('utf-8')
